aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -254,6 +254,157 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
   return at::linalg_matrix_rank(self, c10::nullopt, symmetric);
 }
 
+// multi_dot helper functions
+namespace {
+
+/**
+ * @brief Computes the optimal order to multiply the input tensors
+ *
+ * Follows the dynamic programming algorithm from Cormen et al,
+ * "Introduction to Algorithms, Third Edition", Chapter 15.2,
+ * p. 370-378. Note that the book uses 1-based indexing.
+ *
+ * @param tensors list of 2D tensors
+ * @return 2D vector used by #matrix_chain_multiplication
+ */
+std::vector<std::vector<int64_t>> matrix_chain_order(
+    const std::vector<Tensor>& tensors) {
+  const auto n = tensors.size();
+
+  // Tensor i has dimensions p[i] by p[i + 1]
+  std::vector<int64_t> p(n + 1);
+  for (auto i = decltype(n){0}; i < n; ++i) {
+    p[i] = tensors[i].size(0);
+  }
+  p[n] = tensors[n - 1].size(1);
+
+  // Memoization matrices for the cost and order
+  std::vector<std::vector<int64_t>> m(n, std::vector<int64_t>(n, 0));
+  std::vector<std::vector<int64_t>> s(n, std::vector<int64_t>(n));
+
+  // For each sublist [i, j] consider the cost of splitting the multiplication
+  // into tensors [0...i] and [j...n] and then multiplying the result of those.
+  for (auto l = decltype(n){1}; l < n; ++l) {
+    for (auto i = decltype(n){0}; i < n - l; ++i) {
+      const auto j = i + l;
+      m[i][j] = std::numeric_limits<int64_t>::max();
+      for (auto k = decltype(n){i}; k < j; ++k) {
+        const auto q = m[i][k] + m[k + 1][j] + p[i] * p[k + 1] * p[j + 1];
+        if (q < m[i][j]) {
+          m[i][j] = q;
+          s[i][j] = k;
+        }
+      }
+    }
+  }
+
+  return s;
+}
+
+// Multiplies the tensors following the order from #matrix_chain_order
+Tensor matrix_chain_multiplication(
+    const std::vector<Tensor>& tensors,
+    const std::vector<std::vector<int64_t>>& order,
+    int64_t i,
+    int64_t j) {
+  if (i == j) {
+    return tensors[i];
+  }
+  return at::mm(
+      matrix_chain_multiplication(tensors, order, i, order[i][j]),
+      matrix_chain_multiplication(tensors, order, order[i][j] + 1, j));
+}
+
+} // namespace
+
+Tensor linalg_multi_dot(TensorList _tensors) {
+  const auto n = _tensors.size();
+  TORCH_CHECK(n >= 2, "multi_dot() expected at least 2 tensors but got ", n);
+
+  std::vector<int64_t> out_shape;
+  std::vector<Tensor> tensors(n);
+
+  // If the first tensor is 1D view it as a row vector
+  if (_tensors[0].dim() == 1) {
+    tensors[0] = _tensors[0].view({1, -1});
+  } else if (_tensors[0].dim() == 2) {
+    tensors[0] = _tensors[0];
+    out_shape.push_back(tensors[0].size(0));
```

Comment:
What happens to the out_shape vector when the first or last tensor is 1D? What about when both are 1D?

Reply:
This will be in the docs but basically for first and last tensors with sizes (a, b) x (c, d) the output shape has size (a, d). But if either tensor is 1D, then either a or d is implicitly size 1 but this will be squeezed from the output. So if we have tensors (b) x (c, d) the output shape is (d,).

Reply:
Explained in the docs.